python/ray/train/mosaic/mosaic_trainer.py

```diff
@@ -0,0 +1,168 @@
+        preprocessor: A ray.data.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A MosiacCheckpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        trainer_init_per_worker: Callable[[Optional[Dict]], Trainer],
+        *,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        trainer_init_config: Optional[Dict] = None,
+        torch_config: Optional[TorchConfig] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        dataset_config: Optional[Dict[str, DatasetConfig]] = None,
+        run_config: Optional[RunConfig] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+
+        self._validate_trainer_init_per_worker(
+            trainer_init_per_worker, "trainer_init_per_worker"
+        )
+
+        trainer_init_config = trainer_init_config.copy() if trainer_init_config else {}
+        if "_trainer_init_per_worker" in trainer_init_config:
+            raise ValueError(
+                "'_trainer_init_per_worker' is a reserved key in `trainer_init_config`."
+            )
+        trainer_init_config["_trainer_init_per_worker"] = trainer_init_per_worker
+
+        super().__init__(
+            train_loop_per_worker=_mosaic_train_loop_per_worker,
+            train_loop_config=trainer_init_config,
+            torch_config=torch_config,
+            scaling_config=scaling_config,
+            dataset_config=dataset_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
+
+    def _validate_trainer_init_per_worker(
+        self, trainer_init_per_worker: Callable, fn_name: str
+    ) -> None:
+        num_params = len(inspect.signature(trainer_init_per_worker).parameters)
+        if num_params > 1:
+            raise ValueError(
+                f"{fn_name} should take in at most 1 argument, "
+                f"but it accepts {num_params} arguments instead."
+            )
+
+
+def _mosaic_train_loop_per_worker(config):
+    """Per-worker training loop for Mosaic Composers."""
+    trainer_init_per_worker = config.pop("_trainer_init_per_worker")
+
+    os.environ["RANK"] = str(session.get_world_rank())
+    os.environ["WORLD_SIZE"] = str(session.get_world_size())
+    os.environ["LOCAL_RANK"] = str(session.get_local_rank())
+
+    # get dataset shard -- For now we don't support using ray dataset shards.
+    # the dataloader should be prepared inside the ``trainer_init_per_worker`` function
+    train_dataset = session.get_dataset_shard(TRAIN_DATASET_KEY)
+    eval_dataset = session.get_dataset_shard(EVALUATION_DATASET_KEY)
+
+    # initialize Composer trainer
+    trainer: Trainer = trainer_init_per_worker(**config)
+
+    # call the trainer
+    trainer.fit()
```

Comment:
Does Mosaic take any arguments to `trainer.fit()`?

Reply:
We can add fit config here.

Reply:
`fit_config` has been added as a key for `trainer_init_config`

Reply:
Ah so I'm taking a look at the API: https://docs.mosaicml.com/en/latest/api_reference/generated/composer.Trainer.html#composer.Trainer.fit

and it looks like the `fit` arguments are the same exact as `__init__`. So I think we can actually remove `fit_config` since it's just redundant with `trainer_init_config`.